File: backend/app/engine/plaster.py
"""Plaster / finishes. Unit: m2.

Face area minus openings (> 0.5 m2 per IS 1200). Reveal compensation for
small openings is handled by the no-deduct rule. See project.md section 6.
"""
from __future__ import annotations

from . import materials as mat
from .units import FormulaStep, Quantity, mm_to_m

CLAUSE = "IS 1200 Part 12"


def plaster_surface(m) -> list[Quantity]:
    L, H = mm_to_m(m.length_mm), mm_to_m(m.height_mm)
    n = m.count
    gross = L * H * m.faces

    deduct = 0.0
    for op in m.openings:
        area = mm_to_m(op.width_mm) * mm_to_m(op.height_mm)
        if area > mat.PLASTER_OPENING_DEDUCT_THRESHOLD:   # IS 1200: only > 0.5 m2
            deduct += area * m.faces * op.count

    net = max(gross - deduct, 0.0) * n
    mortar = net * mm_to_m(m.thickness_mm)
    return [
        Quantity(
            category="plaster",
            description=f"Plaster {m.label} ({m.thickness_mm:.0f} mm, {m.faces} face/s)".strip(),
            unit="m2", value=net, nos=n, length_m=L, depth_m=H,
            audit=[FormulaStep(
                "plaster.area",
                "(L*H*faces - openings(>0.5m2))*count",
                {"L_m": L, "H_m": H, "faces": m.faces,
                 "deduct_openings_m2": deduct, "count": n}, net, CLAUSE)],
            extra={"mortar_m3": round(mortar, 3)},
        )
    ]

File: frontend/src/engine/plaster.ts
// Plaster / finishes. Ported from plaster.py. Unit: m2.
import * as mat from "./materials";
import { Member } from "./members";
import { Quantity, mmToM, pyRound, qty, step, fmt0 } from "./units";

const CLAUSE = "IS 1200 Part 12";

export function plaster_surface(m: Member): Quantity[] {
  const L = mmToM(m.length_mm), H = mmToM(m.height_mm);
  const n = m.count;
  const gross = L * H * m.faces;

  let deduct = 0.0;
  for (const op of m.openings || []) {
    const area = mmToM(op.width_mm) * mmToM(op.height_mm);
    if (area > mat.PLASTER_OPENING_DEDUCT_THRESHOLD) deduct += area * m.faces * op.count;
  }

  const net = Math.max(gross - deduct, 0.0) * n;
  const mortar = net * mmToM(m.thickness_mm);
  return [qty({
    category: "plaster",
    description: `Plaster ${m.label} (${fmt0(m.thickness_mm)} mm, ${m.faces} face/s)`.trim(),
    unit: "m2", value: net, nos: n, length_m: L, depth_m: H,
    audit: [step("plaster.area", "(L*H*faces - openings(>0.5m2))*count",
      { L_m: L, H_m: H, faces: m.faces, deduct_openings_m2: deduct, count: n }, net, CLAUSE)],
    extra: { mortar_m3: pyRound(mortar, 3) },
  })];
}
